hazelcast/src/main/java/com/hazelcast/sql/impl/exec/root/BlockingRootResultConsumer.java

```diff
@@ -0,0 +1,194 @@
+    @Override
+    public void setup(QueryFragmentContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public boolean consume(List<Row> batch, boolean last) {
+        assert !batch.isEmpty();
+
+        synchronized (mux) {
+            assert !done;
+
+            if (currentBatch == null) {
+                currentBatch = batch;
+
+                if (last) {
+                    done = true;
+                }
+
+                mux.notifyAll();
+
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    @Override
+    public void onError(HazelcastSqlException error) {
+        synchronized (mux) {
+            if (!done) {
+                done = true;
+                doneError = error;
+
+                mux.notifyAll();
+            }
+        }
+    }
+
+    /**
+     * Poll the next batch from the upstream, waiting if needed.
+     *
+     * @return The batch or {@code null} if end of stream is reached.
+     */
+    private List<Row> awaitNextBatch() {
+        synchronized (mux) {
+            while (true) {
+                // Consume the batch if it is available.
+                if (currentBatch != null) {
+                    List<Row> res = currentBatch;
+
+                    currentBatch = null;
+
+                    return res;
+                }
+
+                // Handle end of the stream.
+                if (done) {
+                    if (doneError != null) {
+                        throw doneError;
+                    }
+
+                    return null;
+                }
+
+                // Otherwise wait.
+                try {
+                    mux.wait();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+
+                    throw HazelcastSqlException.error("Thread was interrupted while waiting for more results.", e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Request the next batch from the executor.
+     */
+    private void requestNextBatch() {
+        synchronized (mux) {
+            if (done) {
+                return;
+            }
+        }
+
+        // We may reach this place only if some rows are already produced, and this is possible only after the setup,
+        // so the context should be initialized.
+        assert context != null;
+
+        context.schedule();
+    }
+
+    @Override
+    public Iterator<Row> iterator() {
+        return iterator;
+    }
+
+    /**
+     * Iterator over results.
+     */
+    private class InternalIterator implements Iterator<Row> {
+
+        private List<Row> batch;
+        private int position;
+
+        @Override
+        public boolean hasNext() {
+            setNextBatchIfNeeded();
+
+            assert batch != null || done;
+
+            return batch != null;
+        }
+
+        @Override
+        public Row next() {
+            setNextBatchIfNeeded();
```

Comment:
That implies `next` can be called without preceding `hasNext` call, is that possible?

Reply:
Yes, it is possible. The `Iterator` interface doesn't require users to call `hasNext` first.

Reply:
I'd replace it with the 
```
if (!hasNext()) {
    throw new NoSuchElementException();
}
```

Reply:
Fixed.